coffeescript.d.ts

```diff
@@ -0,0 +1,247 @@
+import * as babel from "babel-core";
+
+type IObject = Record<string, unknown>;
+
+/**
+ * CoffeeScript compiler options.
+ *
+ * @member ast If true, output an abstract syntax tree of the input CoffeeScript source code.
+ * @member bare If true, omit a top-level IIFE safety wrapper.
+ * @member filename File name to compile - defaults to `index.js`.
+ * @member header If true, output the `Generated by CoffeeScript` header.
+ * @member inlineMap If true, output the source map as a base64-encoded string in a comment at the bottom.
+ * @member sourceMap If true, output a source map object with the code.
+ * @member transpile Babel transpilation options - see `babel.TransformOptions`.
+ */
+export interface ICoffeeScriptOptions {
+  ast?: boolean;
+  bare?: boolean;
+  filename?: string;
+  header?: boolean;
+  inlineMap?: boolean;
+  sourceMap?: boolean;
+  transpile?: babel.TransformOptions;
+}
+
+/**
+ * CoffeeScript abstract syntax tree location data.
+ */
+ export interface CoffeeScriptASTNodeLocationData {
+  first_column: number;
+  first_line: number;
+  last_line: number;
+  last_column: number;
+}
+
+/**
+ * CoffeeScript abstract syntax tree member for base namespaces.
+ */
+ export interface CoffeeScriptASTBase {
+  locationData: CoffeeScriptASTNodeLocationData;
+  value: string;
+}
+
+/**
+ * CoffeeScript abstract syntax tree member for variable names.
+ */
+ export interface CoffeeScriptASTName {
+  locationData: CoffeeScriptASTNodeLocationData;
+  name: CoffeeScriptASTName;
+  soak: boolean;
+}
+
+/**
+ * CoffeeScript abstract syntax tree for properties and indexes.
+ */
+ export interface CoffeeScriptASTProperty {
+  locationData: CoffeeScriptASTNodeLocationData;
+  name: CoffeeScriptASTName;
+}
+
+/**
+ * CoffeeScript's abstract syntax tree node interfaces with all possible node properties.
+ */
+export interface CoffeeScriptASTNode {
+  args: CoffeeScriptASTNode[] | [];
+  base?: CoffeeScriptASTBase;
+  body: CoffeeScriptASTBody;
+  boundFuncs?: CoffeeScriptASTNode[] | [];
+  context?: string;
+  classBody?: boolean;
+  expressions?: CoffeeScriptASTNode[] | [];
+  isNew?: boolean;
+  isSuper?: boolean;
+  locationData: CoffeeScriptASTNodeLocationData;
+  properties?: CoffeeScriptASTProperty[] | [];
+  param?: boolean;
+  parent?: CoffeeScriptASTNode;
+  soak?: boolean;
+  subpattern?: boolean;
+  variable: CoffeeScriptASTNode;
+}
+
+export interface CoffeeScriptASTBody {
+  classBody: boolean;
+  expressions: CoffeeScriptASTNode[] | [];
+  locationData: CoffeeScriptASTNodeLocationData;
+}
+
+/**
+ * List of precompiled CoffeeScript file extensions.
+ */
+export let FILE_EXTENSIONS: string[];
+
+/**
+ * Version number of the CoffeeScript compiler.
+ */
+export let VERSION: string;
+
+/**
+ * Helpers used internally to compile CoffeeScript.
+ *
+ * @deprecated for external use and lacking TypeScript definitions.
+ */
+export interface helpers {
+  [handle: string]: (...args: any[]) => any;
```

Comment:
Why not define each helper?

Reply:
😱 29 helpers, not all documented. I couldn't automate the process and couldn't do it all by hand

Reply:
@GeoffreyBooth done in bd73064